### PR TITLE
[Chef] Implement DoorLock example app

### PR DESCRIPTION
**Pull request by erwinpan1**
Implement more Chef doorlock attributes & command such as:  SetUser, SetCredential, GetWeekDaySchedule, SetWeekDaySchedule ... etc.

### CODE REVIEW

File: examples/chef/common/clusters/door-lock/chef-lock-endpoint.cpp [hunk 1]
```diff
@@ -0,0 +1,679 @@
+/*
+ *
+ *    Copyright (c) 2022-2023 Project CHIP Authors
```
Comment (tcarmelveilleux) [inline examples/chef/common/clusters/door-lock/chef-lock-endpoint.cpp:3]: New file, make copyright dates 2024

File: examples/chef/common/clusters/door-lock/chef-lock-endpoint.cpp [hunk 2]
```diff
@@ -0,0 +1,679 @@
...
+        ChipLogError(Zcl,
+                     "Cannot set user - total number of credentials is too big [endpoint=%d,index=%d,adjustedUserIndex=%u"
+                     ",totalCredentials=%u,maxNumberOfCredentials=%u]",
+                     mEndpointId, userIndex, adjustedUserIndex, static_cast<unsigned int>(totalCredentials),
+                     static_cast<unsigned int>(userInStorage.credentials.capacity()));
+        return false;
+    }
+
+    chip::Platform::CopyString(userInStorage.userName, userName);
+    userInStorage.userName[userName.size()] = 0;
```
Comment (tcarmelveilleux) [inline examples/chef/common/clusters/door-lock/chef-lock-endpoint.cpp:149]: All of the zero-terminating logic is error-prone. Please use std::string for the `userName` to ensure no issues with null-termination. This is your code so you are allowed to do as you wish.

Comment (tcarmelveilleux) [reply examples/chef/common/clusters/door-lock/chef-lock-endpoint.cpp:149]: Actually, the code is using "flat" storage of username :( That means it expects null-termination in storage as well. This is error-prone and dangerous, but would require significant code changes. A better alternative is to keep username size in the structure, separate from the char array, and avoid any use of `strlen()`, and always generate the `CharSpan`s from explicit size kept alongside the string.

File: examples/chef/common/clusters/door-lock/chef-lock-endpoint.cpp (hunk 1 above)
Comment (erwinpan1) [reply examples/chef/common/clusters/door-lock/chef-lock-endpoint.cpp]: Thanks. Fixes done.

File: examples/chef/common/clusters/door-lock/chef-lock-endpoint.cpp (hunk 2 above)
Comment (erwinpan1) [reply examples/chef/common/clusters/door-lock/chef-lock-endpoint.cpp]: Fixed with CharSpan done. PTAL

**Review by tcarmelveilleux — APPROVED** [review]

**Review by lpbeliveau — APPROVED** [review]

File: examples/chef/common/clusters/door-lock/chef-lock-manager.cpp [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+    //    - UserUniqueID MAY be null causing new user record to use 0xFFFFFFFF for UserUniqueID
+    //        otherwise UserUniqueID SHALL be set to the value provided in the new user record
+    //    - UserStatus MAY be null causing new user record to use OccupiedEnabled for UserStatus
+    //        otherwise UserStatus SHALL be set to the value provided in the new user record
+    //   - UserType MAY be null causing new user record to use UnrestrictedUser for UserType
+    //        otherwise UserType SHALL be set to the value provided in the new user record.
+    uint16_t userIndex(1);
+    chip::FabricIndex creator(1);
+    chip::FabricIndex modifier(1);
+    const chip::CharSpan userName = chip::CharSpan::fromCharString("user1"); // default
```
Comment (bzbarsky-apple) [inline examples/chef/common/clusters/door-lock/chef-lock-manager.cpp:118]: In the future, please use "user1"_span or so.

File: examples/chef/common/clusters/door-lock/chef-lock-manager.cpp [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+            Zcl,
+            "Unable to get number of supported holiday schedules when initializing lock endpoint, defaulting to 10 [endpointId=%d]",
+            endpointId);
+        numberOfHolidaySchedules = 10;
+    }
+
+    mEndpoints.emplace_back(endpointId, numberOfSupportedUsers, numberOfSupportedCredentials, numberOfWeekDaySchedulesPerUser,
+                            numberOfYearDaySchedulesPerUser, numberOfCredentialsSupportedPerUser, numberOfHolidaySchedules);
+
+    // Refer to 5.2.10.34. SetUser Command, when Creat a new user record
```
Comment (bzbarsky-apple) [inline examples/chef/common/clusters/door-lock/chef-lock-manager.cpp:105]: "Creat"?

File: examples/chef/common/clusters/door-lock/chef-lock-manager.cpp (hunk 3 above)
Comment (erwinpan1) [reply examples/chef/common/clusters/door-lock/chef-lock-manager.cpp:118]: Got it. Thanks for the advice.

**Pull request merged**